### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by rom1504**
not done yet, please don't merge

### CODE REVIEW

**Comment by rom1504** [comment]
ok it runs now

**Comment by rom1504** [comment]
I'll finish some things tomorrow and then we're good to go

* [ ] basic readme
* [ ] basic tests
* [ ] I discarded the threadpool but actually I'm wondering if we should keep it ?
* [ ] minimal subsampler ?

**Comment by rom1504** [comment]
this brings basically all features of img2dataset so once this merged we can focus on : 
* making download faster
* building cool subsamplers (also maybe rename subsampler to extractor)

File: video2dataset/data_reader.py [hunk 1]
```diff
@@ -57,7 +57,15 @@ def handle_url(url):
...
+class VideoDataReader:
+    def __init__(self) -> None:
+        pass
 
-class Downloader:
-  def __init__(self):
-    pass
+    def __call__(self, row, timeout, retries):
+        key, url = row
+        file_name, file, _ = handle_url(url)
```
Comment (iejMac) [inline video2dataset/data_reader.py:66]: the last return value for handle_url is supposed to be the name you wish to save the video to. So for example sometimes I used this to save YouTube videos as {youtube_id}.ext but I guess maybe we should get rid of it altogether and just commit to key being the save name.

Comment (rom1504) [reply video2dataset/data_reader.py:66]: yeah key should def not be that

but we can totally store the youtube_id and other things in metadata

Comment (iejMac) [reply video2dataset/data_reader.py:66]: it's one of those things that was helpful for debugging but here it's definitely unnecessary, let's remove it

**Comment by rom1504** [comment]
ok I'm going to leave out test for right now, let's add when it's a little further along

**Comment by rom1504** [comment]
so concerns I have after rereading the code today and thinking:
* in img2dataset a shard is small and fast, logging after each shard makes sense. Here maybe not, maybe we want a logger that looks more like https://github.com/rom1504/clip-retrieval/blob/main/clip_retrieval/clip_inference/logger.py

actually that's it

**Comment by rom1504** [comment]
well yes one big thing is we definitely want the subsampler concept right after this PR

**Comment by rom1504** [comment]
ready for reviews

**Comment by iejMac** [comment]
LGTM, checked that it works and left some minor comments.

File: video2dataset/input_sharder.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    - caption_col: the column name of the caption
+    - save_additional_columns: the list of additional columns to save
+    - number_sample_per_shard: the number of samples per shard
+    - done_shards: a set of already done shards
+    """
+
+    def __init__(
+        self,
+        url_list,
+        input_format,
```
[the remark this answers is not in the document]
Comment (rom1504) [reply video2dataset/input_sharder.py:30]: maybe, in img2dataset I found it was fine to have it be explicit

I figure let's keep like that for now, and likely I'll make that input sharder a common package between both slightly later and then add such improvements

File: video2dataset/main.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    elif output_format == "files":
+        sample_writer_class = FilesSampleWriter  # type: ignore
+    elif output_format == "tfrecord":
+        sample_writer_class = TFRecordSampleWriter  # type: ignore
+    elif output_format == "dummy":
+        sample_writer_class = DummySampleWriter  # type: ignore
+    else:
+        raise ValueError(f"Invalid output format {output_format}")
+
+    downloader = Worker(
```
[the remark this answers is not in the document]
Comment (rom1504) [reply video2dataset/main.py:120]: fixed

File: video2dataset/main.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        distributor_fn = multiprocessing_distributor
+    elif distributor == "pyspark":
+        distributor_fn = pyspark_distributor
+    else:
+        raise ValueError(f"Distributor {distributor} not supported")
+
+    distributor_fn(
+        processes_count,
+        downloader,
+        input_sharder,
```
[the remark this answers is not in the document]
Comment (rom1504) [reply video2dataset/main.py:143]: fixed

**Comment by rom1504** [comment]
fixed comments next steps after: 

* logger may be a little non reactive as is, may need improvements
* no subsampler, this is a very important part, need to be introduced right after
* no test, I want to validate a little that it makes sense before adding tests

**Pull request merged**